src/driver/include/driver/workload_parsers.hpp

```diff
@@ -15,16 +15,72 @@
 #ifndef HEADER_0369D27D_9A68_4981_B344_4BAB3EE09A80_INCLUDED
 #define HEADER_0369D27D_9A68_4981_B344_4BAB3EE09A80_INCLUDED
 
+#include <optional>
 #include <boost/filesystem.hpp>
 #include <yaml-cpp/yaml.h>
 
 #include <driver/v1/DefaultDriver.hpp>
 
 namespace genny::driver::v1 {
 
-using YamlParameters = std::map<std::string, YAML::Node>;
 namespace fs = boost::filesystem;
 
+/**
+ * Values stored in a Context are tagged with a type to ensure
+ * they aren't used incorrectly.
+ *
+ * If updating, please add to typeNames in workload_parsers.cpp
+ * Enums don't have good string-conversion or introspection :(
+ */
+enum class Type {
+    kParameter,
+    kActorTemplate,
+    kActorInstance,
+};
+
+/**
+ * Manages scoped context for stored values.
+ *
+ * Only create using a ContextGuard.
+ */
+class Context {
+public:
+    using ContextMap = std::map<std::string, std::pair<YAML::Node, Type>>;
+    Context(Context* enclosing) : _enclosing{enclosing} {}
+
+    std::optional<YAML::Node> get(const std::string&, const Type& type);
+    void insert(const std::string&, const YAML::Node&, const Type& type);
+
+    // Insert all the values of a node, assuming they are of a type.
+    void insert(const YAML::Node&, const Type& type);
+    Context* enclosing() {return _enclosing;}
+
+private:
+    // A context should be a unique_ptr owned by its recursion level.
+    Context* _enclosing;
+    ContextMap _values;
+};
+
+/**
+ * Opens a context, closes it when exiting scope.
+ */
+class ContextGuard {
+public:
+    ContextGuard(Context*& contextPtr) : _contextPtr{contextPtr}, 
+        _managedContext{std::make_unique<Context>(_contextPtr)} {
+
+        _contextPtr = _managedContext.get();
```

Comment:
It took me longer than I care to admit to understand what's going on here. I think that's mostly due to Monday-after-holiday, but I think there could be some slight tweaks to make this rather simple and clever idea easier to grok. Mostly only important if it needs to remain in the header.

It _looks_ weird to have `*&` types, especially since you have the `unique_ptr` manage the ptr-ref that is presumably managed by the caller. Reassigning a ref (especially an member variable) seems potentially quite confusing.

There are cases to use this--and perhaps this is one of those--but I think you could instead have this be something like a linked-list or stack of params (which are the "meat" of the contexts):

```c++
// hpp
class WorkloadParser {
  // lets you defer the impl to the cpp
  std::unique_ptr<typename Context> _context;
  // not a u_p<Ctx*> or u_p<Ctx*&>, just u_p<Ctx>
};

// cpp
struct Context {
  std::stack<std::unordered_map<Node,Type>> _params;
  struct ScopeGuard {
    ScopeGuard() {
      _params.push(...);
    }
    ~ScopeGuard() { _params.pop(); }
  };
  get(k,t) {
     auto top = _params.top();
     while(top) {
       if (k in top.keys()) { return top[k]; }
       top = _params.previous(top); // not sure of the stack api
     }
  }
  insert(k,v,t) {
    _params.top().insert(k,v,t);
  }
  ScopeGuard enter() { return ScopeGuard(); }
};
...
parseExternal() {
  _context->enter();
}
```

To me this is clearer at the expense of not being quite as clever with reassigning refs, but this may not all work out or it may end up less clear than what you started with here. 

ANYWYAY: It seems like this is all "helper" logic for the impl. If that's the case, see if you can put it in the CPP in which case these concerns are, imho, not really all that important as long as the code works (and it seems to according to your tests).

My design-spidey-senses perk up whenever I see things like this in headers which have a tendency to increase in prevalence over time.

Reply:
I followed your stack-based approach; you're right that's a lot clearer. Is there a way to move this into the cpp file but still allow us to unit-test it without the class declaration in the header? I'd prefer to keep the unit tests for `Context`; it's a lot more explicit than the actor-yaml one I added which relies on the parsing behavior of templates and external phases.

Reply:
Cool. And no if we wanna test these impl details specifically then it's gotta be in the header. I sympathize with wanting to keep the test. In that case, please ensure the helper stuff is in the `v1` namespace and/or that it's documented in comments that it's a helper detail and only in the header for ease of testing.